integration-tests/build.gradle.kts

```
@@ -21,7 +21,26 @@ subprojects {
         implementation(project(":integration-tests"))
     }
 
-    val integrationTest by tasks.register<Test>("integrationTest") {
+    /**
+     * Dokka's integration test task is not cacheable because the HTML outputs
+     * it produces when running the tests are used for showcasing resulting documentation.
+     *
+     * At the moment there are two problems that do not allow to make it cacheable:
+     * 1. The task's inputs are such that changes in Dokka's code do not invalidate the cache,
+     *    because it is run with the same version of Dokka ("DOKKA_VERSION") on the same
+     *    test project inputs.
+     * 2. The tests generate HTML output which is then used to showcase documentation.
+     *    The outputs are usually copied to a location from which it will be served.
+     *    However, if the test is cacheable, it produces no outputs, so no documentation
+     *    to showcase. It needs to be broken into two separate tasks: one cacheable for running
+     *    the tests and producing HTML output, and another non-cacheable for copying the output.
+     *
+     * @see [org.jetbrains.dokka.it.TestOutputCopier] for more details on showcasing documentation
+     */
+    @DisableCachingByDefault
+    abstract class NonCacheableIntegrationTest : Test()
+
+    val integrationTest by tasks.register<NonCacheableIntegrationTest>("integrationTest") {
         maxHeapSize = "2G"
         description = "Runs integration tests."
         group = "verification"
```

```suggestion
    val integrationTest by tasks.register<Test>("integrationTest") {
        maxHeapSize = "2G"
        description = "Runs integration tests."
        group = "verification"

        System.getenv("DOKKA_TEST_OUTPUT_PATH")?.let { testOutputDir ->
            outputs.dir(testOutputDir)
        }
```

Hi 👋  If you want the `integrationTest` task to run whenever the HTML directory is not up-to-date, then I think you just need to register the directory as a task output. I presume `DOKKA_TEST_OUTPUT_PATH` holds the the output directory?

---

Hey! Not at all. 

`DOKKA_TEST_OUTPUT_PATH` is not the output path, but it's the path to which the output needs to be copied. The actual test output path is a random temp directory. The problem is that `DOKKA_TEST_OUTPUT_PATH` is almost static (see where it's set in GH workflows), but the temp directory is not. The temp directory is created in the AbstractIntegrationTest somewhere, not in the Gradle task.

To fix it properly, the following result should be achieved:

1. If `DOKKA_TEST_OUTPUT_PATH` is set, the test results need to be copied to that path. To do that, we could cache the test outputs, but the problem here is that on the task level we don't know what the output directory is.
2. If any Dokka code is changed, the task needs to be invalidated. Now it works through publishing Dokka to `mavenLocal`, and the version string is always the same, so it's viewed as the same input and the task is not invalidated properly - it knows nothing of changes in other modules. I'm also not sure if it'll invalidate if any of the test project code changes, as it's also hardcoded in the test itself, I don't think it's set up to be a task input.

All in all, I have a feeling that the current setup is not following best practices, that's why it's causing these problems. Maybe it'll be easier to just refactor it altogether.

You can have a go if you want, but I'd like to merge this in the meantime to make sure the tests are run

---

> Hey! Not at all.
> 
> `DOKKA_TEST_OUTPUT_PATH` is not the output path, but it's the path to which the output needs to be copied. The actual test output path is a random temp directory. The problem is that `DOKKA_TEST_OUTPUT_PATH` is almost static (see where it's set in GH workflows), but the temp directory is not. The temp directory is created in the AbstractIntegrationTest somewhere, not in the Gradle task.

Okay clear, that makes sense. So the order of events is

1. the GitHub action sets `DOKKA_TEST_OUTPUT_PATH`, and triggers the integration test
2. Gradle starts the integration test task...
  1. The Kotlin code runs the tests using `ProcessBuilder()`
  2. HTML is produced to a random temp dir
  3. Finally, HTML is copied from the temp dir to `DOKKA_TEST_OUTPUT_PATH` by `TestOutputCopier`
5. The Gradle task concludes

The problems are on step 2 

- because the integration test projects don't directly depend on the other projects, but instead uses Maven Local, Gradle thinks the source code hasn't changed
- Gradle doesn't realise that the integration tests should output, and so Gradle doesn't realise the outputs aren't up-to-date

So ideally

1. Dokka should use the '[sharing outputs between projects](https://docs.gradle.org/current/userguide/cross_project_publications.html)' config to share the project files, rather than publishing to MavenLocal, so Gradle knows if the source code has changed
2. Somehow make the HTML output stable (there are lots of ways to do this - split up the integration tests, set stable directories via System Properties so Gradle is aware of them, use a single directory...)


The one other related improvement I can think of is to convert the AWS sync shell script `aws_sync.sh` into a Gradle task, with properly registered task inputs/outputs.



> You can have a go if you want, but I'd like to merge this in the meantime to make sure the tests are run

Yes that make sense - skipping the build cache for the integration tests is a good fix.

---

Yeah, I think you got it right :+1: I'm not sure what the best solution/implementation for these tests should be, it needs to be researched more, but I do have a strong feeling it's not done right at the moment :sweat_smile: